### PR TITLE
Fixes to speed up running queries — add caching & avoid work

### PR DESCRIPTION
**Pull request by KyleAMathews**

### CODE REVIEW

File: packages/gatsby-plugin-sharp/src/index.js [hunk 1]
```diff
@@ -11,6 +11,19 @@ const imageminWebp = require(`imagemin-webp`)
 const queue = require(`async/queue`)
 const path = require(`path`)
 
+const imageSizeCache = new Map()
```
Comment (KyleAMathews) [inline packages/gatsby-plugin-sharp/src/index.js]: This cut ~700ms off a number of queries on gatsbyjs.org

File: packages/gatsby/src/internal-plugins/query-runner/page-query-runner.js [hunk 2]
```diff
@@ -120,19 +120,21 @@ const runQueriesForIds = ids => {
 
 const findDirtyIds = actions => {
   const state = store.getState()
-  return actions.reduce((dirtyIds, action) => {
-    const node = action.payload
+  return _.uniq(
```
Comment (KyleAMathews) [inline packages/gatsby/src/internal-plugins/query-runner/page-query-runner.js]: There were a ton of duplicates (though this wasn't too big of deal as our query queue removes duplicates).

File: packages/gatsby/src/redux/reducers/components.js [hunk 3]
```diff
@@ -9,11 +9,10 @@ module.exports = (state = {}, action) => {
     case `CREATE_LAYOUT`:
       action.payload.componentPath = normalize(action.payload.component)
       state[action.payload.componentPath] = _.merge(
-        {},
+        { query: `` },
```
Comment (KyleAMathews) [inline packages/gatsby/src/redux/reducers/components.js]: sometimes we were overwriting the query so it'd stop running 🤦‍♂️

File: packages/gatsby/src/schema/node-tracking.js [hunk 4]
```diff
@@ -29,14 +29,21 @@ const addRootNodeToInlineObject = (data, nodeId) => {
  * and that Node object.
  * @param {Node} node Root Node
  */
+const nodeDigestTracked = new Set()
 const trackInlineObjectsInRootNode = node => {
+  if (nodeDigestTracked.has(node.internal.contentDigest)) {
```
Comment (KyleAMathews) [inline packages/gatsby/src/schema/node-tracking.js]: This can get quite expensive, caching helps

**Comment by KyleAMathews** [comment]
This offset some of the build time regressions of late but not entirely.

First thing after v2 I'll be back to working on getting a new testing infrastructure built — which will include build performance testing so we'll catch regressions easier.

**Comment by KyleAMathews** [comment]
Deploy preview for *using-glamor* failed.

Built with commit cca7abf9e5ccbae495954571d9a6985f694daad2

https://app.netlify.com/sites/using-glamor/deploys/5a90a3e048769b6f358ea975

File: packages/gatsby/src/schema/node-tracking.js [hunk 5]
```diff
@@ -29,14 +29,25 @@ const addRootNodeToInlineObject = (data, nodeId) => {
  * and that Node object.
  * @param {Node} node Root Node
  */
+// const nodeDigestTracked = new Set()
 const trackInlineObjectsInRootNode = node => {
+  // const id =
```
Comment (KyleAMathews) [inline packages/gatsby/src/schema/node-tracking.js]: This broke tests...

File: packages/gatsby-plugin-sharp/src/index.js (hunk 1 above)
Comment (fk) [reply packages/gatsby-plugin-sharp/src/index.js]: 👍 👍

**Pull request merged**